---
title: "MRC_CSC_Sequencing_Process"
author: "MRC CSC"
date: "11/03/2015"
output: html_document
---

This report is generated as part of the Medical Research Council Clinical Sciences Centre Bioinformatics Team's illumina sequencing pipeline.

This report covers basic sequencing results including Illumina sequencing statistics, multiplexing numbers and FASTQC.


Be nice to produce a sam header with this information..to add to aligned files later on.

```{r,echo=F,message=T,warning=F}

split_path <- function(path) {
  if (dirname(path) %in% c(".", path)) return(basename(path))
  return(c(basename(path), split_path(dirname(path))))
}

## Read in config file.
## The config file contains paths to directories and programs.

library(raster)
library(XML)
config = data.frame(readIniFile("~/genomicsProcessing/config.ini"))

## Now look in folders to identify Runs to be processed

basedir <- as.vector(config[config$section == "paths" & config$name == "basedir","value"])

if(dir.exists(basedir)){
  message("Searching ",basedir," for unprocessed runs ..",appendLF = T)
}else{
  stop(basedir," does not exist.\n Stopping processing")
}
subFoldersFull <- dir(file.path(basedir,""),recursive=F,include.dirs=T,full.names=T)
subFoldersFull <- gsub("/+","/",subFoldersFull)
runsToExclude <- as.vector(read.delim(as.vector(config[config$section == "paths" & config$name == "runsToExclude","value"]),h=F)[,1])
subFoldersFull <- subFoldersFull[!basename(subFoldersFull) %in% basename(runsToExclude)]
subFoldersFull <- subFoldersFull[dir.exists(subFoldersFull)]
subFoldersFull_RelativePaths <- gsub(paste0(basedir,"\\/."),"",subFoldersFull)

## Find folders containing RTAComplete.txt, Unaligned and Unaligned/Demultiplexing.html
#total_Run_folders <- length(unique(lapply(subFoldersFull_RelativePaths,function(x) split_path(x)[[length(split_path(x))]])))
total_Run_folders <- length(subFoldersFull)
Run_folders_WithRTA <- subFoldersFull_RelativePaths[file.exists(file.path(subFoldersFull,"RTAComplete.txt"))]
# Run_folders_WithUnaligned <- subFoldersFull_RelativePaths[dir.exists(file.path(subFoldersFull,"Unaligned"))
#                                                           & dir.exists(subFoldersFull)
#                                                             ]
# Run_folders_WithDemultiplexingComplete <- subFoldersFull[file.exists(file.path(subFoldersFull,"Unaligned","Demultiplexing\\.html"))]

# RunsComplete <- Run_folders_WithRTA
# RunsComplete_Processing <- intersect(Run_folders_WithRTA,Run_folders_WithUnaligned)
# RunsComplete_Processing_Demultiplexed <- intersect(RunsComplete_Processing,Run_folders_WithDemultiplexingComplete)
# RunsComplete_Processing_To_Be_Demultiplexed <- setdiff(RunsComplete_Processing,RunsComplete_Processing_Demultiplexed)
# 
# knitr:::kable(data.frame(Total_Runs=total_Run_folders,
#                          Total_Complete_Runs=length(RunsComplete),
#            Total_Demuliplexed_Runs=length(RunsComplete_Processing_Demultiplexed),
#            Total_Awaiting_Demultiplexing=length(RunsComplete_Processing_To_Be_Demultiplexed)))

#foldersFull <- dir(file.path(basedir,""),recursive=F,include.dirs=T,full.names=T)
#foldersFull_RelativePaths <- gsub(paste0(basedir,"\\/."),"",foldersFull)

runParams <- vector("list",length=length(Run_folders_WithRTA))
shellBCLs <- vector()
p <- 1

for(i in 1:length(Run_folders_WithRTA)){
  currentRun <- subFoldersFull[grepl(Run_folders_WithRTA[i],subFoldersFull)]
  xmlFromPresentRunFolder <- xmlParse(file.path(currentRun,"runParameters.xml"))
  currentRunParameters <- xmlToDataFrame(xmlFromPresentRunFolder)
  currentRunParameters <- currentRunParameters[!is.na(currentRunParameters$ExperimentName),,drop=F]
  sampleSheetName <- file.path(currentRun,paste0(currentRunParameters$Barcode,".csv"))
  if(file.exists(sampleSheetName)){
    ss <- read.delim(sampleSheetName,sep=",",quote=NULL,header=T,stringsAsFactors=F)
    ss$SampleID <- gsub("^X","Sample_",validNames(ss$SampleID))
    ss$SampleID <-gsub("\\?|\\(|\\)|\\[|\\]|\\\\|/|\\=|\\+|<|>|\\:|\\;|\"|\'|\\*|\\^|\\||\\&|\\.","_",ss$SampleID)    
    ss$Index <- gsub(" ","",ss$Index)
    toBeTrimmed <- which(lapply(ss$Index,nchar) > as.numeric(currentRunParameters$IndexRead1))
    for(i in toBeTrimmed){
      ss$Index[toBeTrimmed] <-  substr(ss$Index[toBeTrimmed],0,as.numeric(currentRunParameters$IndexRead1))
    }
    
    message("Read samplesheet ",basename(sampleSheetName)," discovered for run ",currentRun)
    index1Lengths <- unlist(lapply(ss$Index,function(x)nchar(x)))
    
    if(any(colnames(ss) %in% "Index2")){
    ss$Index2 <- gsub(" ","",ss$Index2)
    index2Lengths <- unlist(lapply(ss$Index2,function(x)nchar(x)))
    index2NAs <- unlist(lapply(ss$Index2,function(x)is.na(x)))
    index2Lengths[index2NAs] <- 0
    allIndexTypes <- paste0(index1Lengths,"-",index2Lengths)
    uniqueIndexTypes <- unique(allIndexTypes)
    #ss$SampleID <- gsub("[[:punct:]]", "_", ss$SampleID).)
    #ss$SampleID <- gsub("[^[:alnum:]]", "_", ss$SampleID).)
    ss$Index <- gsub("-NA|-$|^-$","",paste(ss$Index,ss$Index2,sep="-"))
    ss <- ss[,-grep("Index2",colnames(ss))]
    }else{
      allIndexTypes <- paste0(index1Lengths)
      uniqueIndexTypes <- unique(allIndexTypes)
      #ss$SampleID <- gsub("[[:punct:]]", "_", ss$SampleID).)
      #ss$SampleID <- gsub("[^[:alnum:]]", "_", ss$SampleID).)
    }
 
    for(l in uniqueIndexTypes){
      tempss <- ss[allIndexTypes %in% l,]
      tempss[is.na(tempss)] <- ""
      if(!ncol(tempss) > 9){
        
      }
      if(!file.exists(gsub("\\.csv",paste0("_",l,"\\.csv"),sampleSheetName))){
        write.table(tempss,file=gsub("\\.csv",paste0("_",l,"\\.csv"),sampleSheetName),quote=F,sep=",",row.names=F,col.names=T)
      }
      #if(!dir.exists(gsub("\\.csv",paste0("_",l),sampleSheetName))){
      #  dir.create(gsub("\\.csv",paste0("_",l),sampleSheetName),showWarnings = F)
      #}
      if(!file.exists(gsub("\\.csv",paste0("_",l,"\\.sh"),sampleSheetName))){
        indexLengths <- unlist(strsplit(l,"-"))
        if(as.numeric(as.vector(currentRunParameters$IndexRead1)) != 0){
          numberOfNs <- as.numeric(as.vector(currentRunParameters$IndexRead1)) - as.numeric(indexLengths[1])
          if(indexLengths[1] == 0){ 
            indexMask1 <- ""
          }else{
            indexMask1 <-  paste0(",I",indexLengths[1],paste0(rep("n",numberOfNs),collapse=""))
          }
        }else{
          indexMask1 <- ""
        }
        if(as.numeric(as.vector(currentRunParameters$IndexRead2)) != 0){
          numberOfNs <- as.numeric(as.vector(currentRunParameters$IndexRead2)) - as.numeric(indexLengths[2])
          if(indexLengths[2] == 0){ 
            indexMask2 <- ""
          }else{
          indexMask2 <-  paste0(",I",indexLengths[2],paste0(rep("n",numberOfNs),collapse=""))
          }
        }else{
          indexMask2 <- ""
        }
        
        usebasemask <- paste0("y*n",indexMask1,indexMask2,",y*n")
        
      runBCLcommand <- paste0(
        as.vector(config[config$section == "programs" & config$name == "configureBclToFastq","value"]),
        " --input-dir ",
        file.path(currentRun,as.vector(config[config$section == "paths" & config$name == "inputdir","value"])),
        " --sample-sheet ",
        gsub("\\.csv",paste0("_",l,"\\.csv"),sampleSheetName),
        " --fastq-cluster-count=0 --use-bases-mask ",usebasemask
        ,
        " --output-dir ",
        gsub("\\.csv",paste0("_",l),sampleSheetName),
        "\n"
        )
      cat(runBCLcommand,file=gsub("\\.csv",paste0("_",l,"\\.sh"),sampleSheetName))
      shellBCLs[p] <- gsub("\\.csv",paste0("_",l,"\\.sh"),sampleSheetName)
      p <- p+1
    }
    }
  }else{
    stop("No samplesheet ",basename(sampleSheetName)," discovered for run ",basename(currentRun))
  }
}


####
#### "FCID,Lane,SampleID,SampleRef,Index,Description,Control,Recipe,Operator,SampleProject\n";

blcOuts <- vector("list",length=length(shellBCLs))
for(i in 1:length(shellBCLs)){
  #blcOuts[[i]] <- system(command=paste0("bash ",shellBCLs[i]),intern=T)
  blcOuts[[i]] <- system2("bash",shellBCLs[i],stdout=T,stderr=T)
}

makeQsubs <- vector("list",length=length(shellBCLs))
for(i in 1:length(shellBCLs)){

makeQsubs[i] <- paste0("qsub -v PATH -wd ",
gsub("\\.sh","/",shellBCLs[i]),
" -o ",
gsub("\\.sh","\\.out",shellBCLs[i]),
" -e ", 
gsub("\\.sh","\\.error",shellBCLs[i]),
" -N ",
basename(gsub("\\.sh","",shellBCLs[i])),
" -pe make -10 /ifs/data/Bioinformatics/Scripts/Casava_demultiplex.sh"
)
}

makeOuts <- vector("list",length=length(shellBCLs))
for(i in 1:length(makeQsubs)){
  #blcOuts[[i]] <- system(command=paste0("bash ",shellBCLs[i]),intern=T)
  makeOuts[[i]] <- system(makeQsubs[[i]],intern=T)
  cat(makeQsubs[[i]],file=paste0(gsub(" ","_",Sys.time()),"_Qsubs.sh\n"),append=T)
}

nohup make -j 10 & -- Works
  
qsub -wd /ifs/data/Hiseq/Runs/151113_D00467_0149_BC7T3UANXX/C7J3UANXX_8-8 -o /ifs/data/Hiseq/Runs/151113_D00467_0149_BC7T3UANXX/C7J3UANXX_8_8.out -e /ifs/data/Hiseq/Runs/151113_D00467_0149_BC7T3UANXX/C7J3UANXX_8_8.error -n C7J3UANXX_8-8  -pe make -4 /ifs/data/Bioinformatics/Scripts/Casava_demultiplex.sh

```

