---
title: "MRC_CSC_Sequencing_Process"
author: "MRC CSC"
date: "11/03/2015"
output: html_document
---

This report is generated as part of the Medical Research Council Clinical Sciences Centre Bioinformatics Team's illumina sequencing pipeline.

This report covers basic sequencing results including Illumina sequencing statistics, multiplexing numbers and FASTQC.


Be nice to produce a sam header with this information..to add to aligned files later on.

```{r,echo=F,message=T,warning=F}

split_path <- function(path) {
  if (dirname(path) %in% c(".", path)) return(basename(path))
  return(c(basename(path), split_path(dirname(path))))
}

## Read in config file.
## The config file contains paths to directories and programs.

library(raster)
library(XML)
config = data.frame(readIniFile("config.ini"))

## Now look in folders to identify Runs to be processed

basedir <- as.vector(config[config$section == "paths" & config$name == "basedir","value"])

if(dir.exists(basedir)){
  message("Searching ",basedir," for unprocessed runs ..",appendLF = T)
}else{
  stop(basedir," does not exist.\n Stopping processing")
}
subFoldersFull <- dir(file.path(basedir,""),recursive=F,include.dirs=T,full.names=T)
subFoldersFull_RelativePaths <- gsub(paste0(basedir,"\\/."),"",subFoldersFull)


## Find folders containing RTAComplete.txt, Unaligned and Unaligned/Demultiplexing.html
#total_Run_folders <- length(unique(lapply(subFoldersFull_RelativePaths,function(x) split_path(x)[[length(split_path(x))]])))
total_Run_folders <- length(subFoldersFull)
Run_folders_WithRTA <- subFoldersFull_RelativePaths[file.exists(file.path(subFoldersFull,"RTAComplete.txt"))]
Run_folders_WithUnaligned <- subFoldersFull_RelativePaths[dir.exists(file.path(subFoldersFull,"Unaligned"))
                                                          & dir.exists(subFoldersFull)
                                                            ]
Run_folders_WithDemultiplexingComplete <- subFoldersFull_RelativePaths[grep("Demultiplexing\\.html",subFoldersFull_RelativePaths)]

RunsComplete <- dirname(Run_folders_WithRTA)
RunsComplete_Processing <- intersect(dirname(Run_folders_WithRTA),dirname(Run_folders_WithUnaligned))
RunsComplete_Processing_Demultiplexed <- intersect(RunsComplete_Processing,rev(split_path(Run_folders_WithDemultiplexingComplete))[[1]])
RunsComplete_Processing_To_Be_Demultiplexed <- setdiff(RunsComplete_Processing,RunsComplete_Processing_Demultiplexed)

knitr:::kable(data.frame(Total_Runs=total_Run_folders,
                         Total_Complete_Runs=length(RunsComplete),
           Total_Demuliplexed_Runs=length(RunsComplete_Processing_Demultiplexed),
           Total_Awaiting_Demultiplexing=length(RunsComplete_Processing_To_Be_Demultiplexed)))

foldersFull <- dir(file.path(basedir,""),recursive=F,include.dirs=T,full.names=T)
foldersFull_RelativePaths <- gsub(paste0(basedir,"\\/."),"",foldersFull)

runParams <- vector("list",length=length(RunsComplete_Processing_To_Be_Demultiplexed))


for(i in 1:length(RunsComplete_Processing_To_Be_Demultiplexed)){
  currentRun <- foldersFull[grepl(RunsComplete_Processing_To_Be_Demultiplexed,foldersFull_RelativePaths)]
  xmlFromPresentRunFolder <- xmlParse(file.path(currentRun,"runParameters.xml"))
  currentRunParameters <- xmlToDataFrame(xmlFromPresentRunFolder)
  currentRunParameters <- currentRunParameters[!is.na(currentRunParameters$ExperimentName),,drop=F]
  sampleSheetName <- file.path(currentRun,paste0(currentRunParameters$Barcode,".csv"))
  if(file.exists(sampleSheetName)){
    ss <- read.delim(sampleSheetName,sep=",",quote=NULL,header=T,stringsAsFactors=F)
    message("Read samplesheet ",basename(sampleSheetName)," discovered for run ",currentRun)
    index1Lengths <- unlist(lapply(ss$Index,function(x)nchar(x)))
    index2Lengths <- unlist(lapply(ss$Index2,function(x)nchar(x)))
    index2NAs <- unlist(lapply(ss$Index2,function(x)is.na(x)))
    index2Lengths[index2NAs] <- 0
    allIndexTypes <- paste0(index1Lengths,"-",index2Lengths)
    uniqueIndexTypes <- unique(allIndexTypes)
    #ss$SampleID <- gsub("[[:punct:]]", "_", ss$SampleID).)
    #ss$SampleID <- gsub("[^[:alnum:]]", "_", ss$SampleID).)
    ss$SampleID <- gsub("^X","Sample_",validNames(ss$SampleID))
    ss$SampleID <-gsub("\\?|\\(|\\)|\\[|\\]|\\\\|/|\\=|\\+|<|>|\\:|\\;|\"|\'|\\*|\\^|\\||\\&|\\.","_",ss$SampleID)
    for(l in uniqueIndexTypes){
      tempss <- ss[allIndexTypes %in% l,]
      tempss[is.na(tempss)] <- ""
      write.table(tempss,file=gsub("\\.csv",paste0("_",l,"\\.csv"),sampleSheetName),quote=F,sep=",",row.names=F,col.names=T)
      if(!dir.exists(gsub("\\.csv",paste0("_",l),sampleSheetName))){
        dir.create(gsub("\\.csv",paste0("_",l),sampleSheetName),showWarnings = F)
      }
      runBCLcommand <- paste0(
        as.vector(config[config$section == "programs" & config$name == "configureBclToFastq","value"]),
        " --input-dir ",
        file.path(currentRun,as.vector(config[config$section == "paths" & config$name == "inputdir","value"])),
        " --sample-sheet ",
        gsub("\\.csv",paste0("_",l,"\\.csv"),sampleSheetName),
        " --fastq-cluster-count=0 --use-bases-mask ",
        paste0("y*n","I",currentRunParameters$IndexRead1,"n","y*n"),
        " --output-dir",
        gsub("\\.csv",paste0("_",l),sampleSheetName),
        "\n"
        )
      cat(runBCLcommand,file=gsub("\\.csv",paste0("_",l,"\\.sh"),sampleSheetName))
    }
  }else{
    stop("No samplesheet ",basename(sampleSheetName)," discovered for run ",basename(currentRun))
  }
}




```

